### PR TITLE
[SPARK-29000][SQL] Decimal precision overflow when don't allow precision loss

### PR DESCRIPTION
**Pull request by turboFei**
### What changes were proposed in this pull request?

When we set spark.sql.decimalOperations.allowPrecisionLoss to false.

For the sql below, the result will overflow and return null.

Case a:

`select case when 1=2 then 1 else 1.000000000000000000000001 end * 1`
Similar  with the division operation.

This sql below will lost precision.

Case b:

`select case when 1=2 then 1 else 1.000000000000000000000001 end / 1`

Let us check the code of TypeCoercion.scala.

 https://github.com/apache/spark/blob/a75467432eab2ee38b93906d86416fe9e110e82e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L864-L875.

For binaryOperator, if the two operands have differnt datatype, rule ImplicitTypeCasts will find a  common type and cast both operands to common type.

So, for these cases menthioned,  their left operand is Decimal(34, 24) and right operand is Literal.

Their common type is Decimal(34,24), and Literal(1) will be casted to Decimal(34,24).

Then both operands are decimal type and they will be processed by decimalAndDecimal method of DecimalPrecision class.

Let's check the relative code.

https://github.com/apache/spark/blob/a75467432eab2ee38b93906d86416fe9e110e82e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala#L123-L153

When we don't allow precision loss, the result type of multiply operation in case a is Decimal(38, 38), and that of division operation in case b is Decimal(38, 20).

Then the multi operation in case a will overflow and division operation in case b will lost precision.

In this PR, we skip to handle the  binaryOperator if DecimalType operands are involved and rule `DecimalPrecision` will handle it.

### Why are the changes needed?

Data will corrupt without this change.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by turboFei** [comment]
@cloud-fan @dongjoon-hyun @gatorsmile  I saw you worked on this files. Maybe you can help reviewing the PR.  Thanks.

**Comment by maropu** [comment]
also cc: @mgaido91

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala [hunk 1]
```diff
@@ -1224,6 +1224,15 @@ class HiveQuerySuite extends HiveComparisonTest with SQLTestUtils with BeforeAnd
...
   }
+
+  test("SPARK-29000: arithmetic computation overflow when don't allow decimal precision loss ") {
+    val allow = sql("set spark.sql.decimalOperations.allowPrecisionLoss")
+    val originValue = allow.collect().apply(0).apply(1)
+    sql("set spark.sql.decimalOperations.allowPrecisionLoss=false")
+    val df = sql("select case when 1=2 then 1 else 100.000000000000000000000000 end * 1")
+    QueryTest.checkAnswer(df, Array(Row(100)))
+    sql(s"set spark.sql.decimalOperations.allowPrecisionLoss=$originValue")
+  }
```
Comment (maropu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: Why did you add tests in package `hive`? I think this issue is not only for hive (I didn't carefully check yet)

Comment (turboFei) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: I have moved it into SQLQuerySuite in  package `sql`. Thanks.

Comment (maropu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: I think we need tests in `TypeCoercionSuite`, too.

Comment (turboFei) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala]: I have added relative tests in TypeCoericionSuite.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -3180,6 +3180,15 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession {
     }
 
   }
+
+  test("SPARK-29000: arithmetic computation overflow when don't allow decimal precision loss ") {
+    val allow = sql("set spark.sql.decimalOperations.allowPrecisionLoss")
```
Comment (mgaido91) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: please use `withSQLConf`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 3]
```diff
@@ -862,17 +863,33 @@ object TypeCoercion {
...
-            val newLeft = if (left.dataType == commonType) left else Cast(left, commonType)
-            val newRight = if (right.dataType == commonType) right else Cast(right, commonType)
-            b.withNewChildren(Seq(newLeft, newRight))
-          } else {
-            // Otherwise, don't do anything with the expression.
-            b
-          }
-        }.getOrElse(b)  // If there is no applicable conversion, leave expression unchanged.
+        (left, right) match {
+          // this process logic is a copy of nondecimalAndDecimal method of DecimalPrecision class
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: can we avoid the copy and reuse it then?

**Review by mgaido91 — COMMENTED** [review]
@turboFei thanks for this PR. I don't understand very well the difference before and after your patch. Indeed, it seems to be that the case when operand always becomes a `decimal(34, 24)` as well as the `1` or `1.0`. May you please explain what changes before and after?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 2 above)
Comment (turboFei) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: thanks

**Comment by turboFei** [comment]
> @turboFei thanks for this PR. I don't understand very well the difference before and after your patch. Indeed, it seems to be that the case when operand always becomes a `decimal(34, 24)` as well as the `1` or `1.0`. May you please explain what changes before and after?

We can check the code  below:
https://github.com/apache/spark/blob/4664a082c2c7ac989e818958c465c72833d3ccfe/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala#L123-L128

If  a decimal(34,24) * 1.0, p1=34, s1=24, p2=2,s2=1,
And the result type  is Decimal(38,25)
image-20190905145151701

If a decimal(34,24) * 1,  p1 =34,s1=24,p2=34,s2=24.
And the result type is Decimal(38,38).
image-20190905145246902

The result will overflow.

**Comment by mgaido91** [comment]
so actually my understanding is that the problem is in the order rules are applied. Indeed, in `typeCoercionRules`, there are several rules between `DecimalPrecision` and `ImplicitTypeCasts` and if any of the rules in between (as in this case `CaseWhenCoercion` does) change the return type of one operand, this issue may arise. We might also simply skip to handle decimals in `ImplicitTypeCasts`, waiting for the next pass of rules to be applied so that `DecimalPrecision` handles them.

**Comment by turboFei** [comment]
Yes, I think we should skip to handle decimals in ImplictTypeCasts too.
I wanted reuse all the logic in nondecimalAndDecimal method originally.
However, do you think the operation that cast decimal to double will loss precision?
https://github.com/apache/spark/blob/4664a082c2c7ac989e818958c465c72833d3ccfe/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DecimalPrecision.scala#L326-L329

**Comment by turboFei** [comment]
cc @cloud-fan

**Comment by mgaido91** [comment]
> However, do you think the operation that cast decimal to double will loss precision?

I can cause precision loss as double allows 15-17 significant digits to be correctly represented. But the other way around we may cause overflow, so....

**Comment by turboFei** [comment]
I have updated PR. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 4]
```diff
@@ -862,17 +862,23 @@ object TypeCoercion {
...
-            val newRight = if (right.dataType == commonType) right else Cast(right, commonType)
-            b.withNewChildren(Seq(newLeft, newRight))
-          } else {
-            // Otherwise, don't do anything with the expression.
-            b
-          }
-        }.getOrElse(b)  // If there is no applicable conversion, leave expression unchanged.
+        (left, right) match {
+          // Skip to handle decimals
+          case (l, r) if l.dataType.isInstanceOf[DecimalType] ||
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: can't we just add these checks to the case above?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 5]
```diff
@@ -3180,6 +3180,16 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession {
     }
 
   }
+
+  test("SPARK-29000: arithmetic computation overflow when don't allow decimal precision loss ") {
+    withSQLConf(("spark.sql.decimalOperations.allowPrecisionLoss", "false")) {
```
Comment (mgaido91) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: please use `SQLConf.....key -> "false"`

Comment (turboFei) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 4 above)
Comment (turboFei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 6]
```diff
@@ -861,7 +861,8 @@ object TypeCoercion {
       // Skip nodes who's children have not been resolved yet.
       case e if !e.childrenResolved => e
 
-      case b @ BinaryOperator(left, right) if left.dataType != right.dataType =>
+      case b @ BinaryOperator(left, right) if !left.dataType.isInstanceOf[DecimalType] &&
+        !right.dataType.isInstanceOf[DecimalType] && left.dataType != right.dataType =>
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: nit: 2 spaces indentation more

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala [hunk 7]
```diff
@@ -1576,6 +1576,20 @@ class TypeCoercionSuite extends AnalysisTest {
         SpecifiedWindowFrame(RangeFrame, CurrentRow, UnboundedFollowing))
     )
   }
+
+  test("SPARK-29000 skip to handle decimals in ImplicitTypeCasts") {
```
Comment (mgaido91) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]:
```suggestion
  test("SPARK-29000: skip to handle decimals in ImplicitTypeCasts") {
```

**Review by mgaido91 — COMMENTED** [review]
the change seems reasonable to me, may a committer start a build for this?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala (hunk 7 above)
Comment (turboFei) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 6 above)
Comment (turboFei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: thanks

**Comment by turboFei** [comment]
@maropu  Hi, Can you help me start a build for this PR?

**Comment by maropu** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 8]
```diff
@@ -861,7 +861,8 @@ object TypeCoercion {
       // Skip nodes who's children have not been resolved yet.
       case e if !e.childrenResolved => e
 
-      case b @ BinaryOperator(left, right) if left.dataType != right.dataType =>
+      case b @ BinaryOperator(left, right) if !left.dataType.isInstanceOf[DecimalType] &&
+          !right.dataType.isInstanceOf[DecimalType] && left.dataType != right.dataType =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: can we add a comment to say that it will be handled by another rule `DecimalPrecision`?

Comment (turboFei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: thanks

**Comment by turboFei** [comment]
Jenkins failed on ObjectHashAggregateSuite. But I can pass this suite locally.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 9]
```diff
@@ -3180,6 +3180,18 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession {
     }
 
   }
+
+  test("SPARK-29000: arithmetic computation overflow when don't allow decimal precision loss ") {
+    withSQLConf("spark.sql.decimalOperations.allowPrecisionLoss" -> "false") {
```
Comment (mgaido91) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]:
```suggestion
    withSQLConf(SQLConf.DECIMAL_OPERATIONS_ALLOW_PREC_LOSS.key -> "false") {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 10]
```diff
@@ -861,7 +861,9 @@ object TypeCoercion {
       // Skip nodes who's children have not been resolved yet.
       case e if !e.childrenResolved => e
 
-      case b @ BinaryOperator(left, right) if left.dataType != right.dataType =>
+      // For DecimalType, it will be handled by another rule DecimalPrecision.
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]:
```suggestion
      // If DecimalType operands are involved, DecimalPrecision will handle it
```

**Review by mgaido91 — COMMENTED** [review]
LGTM apart minor comments. Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 9 above)
Comment (turboFei) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 10 above)
Comment (turboFei) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: thanks

**Comment by turboFei** [comment]
Test passed. @maropu  @cloud-fan

**Comment by maropu** [comment]
The code and tests looks ok, but the PR description is ambiguous. Could you brush up it more clearly? what's the root cause and how to fix it, brabrabra...

**Comment by turboFei** [comment]
@maropu  Thanks for you suggestion, I have updated the PR description. Can you help take a look again? Thanks in advance!

**Review by maropu — APPROVED** [review]
LGTM @cloud-fan

**Comment by cloud-fan** [comment]
thanks, merging to master!
